### ISSUE TITLE
support for asciidoctor

### ISSUE THREAD

**Issue by zaxebo1** [ashes999/butterfly#7]
please support asciidoctor too (see http://www.asciidoctor.org). It is sane syntax, and very popular too. It is written in ruby. Probably somekind of additional plugin into butterfly will be able to do this.
support for asciidoctor will get me converted to butterfly.

 Groovy language documentation, spring framework documentation all are in asciidoc. Github.com, bitbucket etc all support asciidoctor natively

**Comment by ashes999** [comment]
Hi @zaxebo1,

Unfortunately, I don't think I can do this. There are a couple of reasons why:

1) AsciiDoctor is written in Ruby. Butterfly is written in Haxe. Haxe => Ruby works, but the other way is much more difficult -- probably a rewrite. That might be worth it, but for reason 2.
2) AsciiDoctor isn't as popular as Markdown -- probably not by a long shot. GitHub wrote [a nice blog post here](https://github.com/blog/2136-a-look-behind-our-decision-to-standardize-on-a-single-markdown-engine-for-github-pages) about why they stopped supporting different variations of Markdown. (TLDR: less code is less complexity means more time spent on improvements; and it doesn't make sense to support it for a marginal number of users.)

It doesn't make sense for me to rewrite AsciiDoctor for Butterfly. But, if you have an existing site that uses AsciiDoctor, making a converter to Butterfly might make sense.

**Comment by ashes999** [comment]
Closing this since three days passed without a response. Please re-open it if there's something more you want to do here (eg. discuss building a site converter).

**Comment by zaxebo1** [comment]
Sorry for late response. i was out of station, (and again will be for one week) so could not respond.

you need not rewrite asciidoctor for butterfly. Just a few wrapper lines in haxe  over asciidoctor, should turn it  into plugin for your butterfly.

for example:    As "restructured text(rst)" markup format compiler is fully written in "python" and "jekyll" static website software is written in "ruby". Hence this small jekyll-rst plugin ( https://github.com/xdissent/jekyll-rst  ) written in ruby and a bit of python, enables one to use jekyll with rst

for example: this following link shows that python based "nikola" static website generator can invoke PHP based markups and even pandoc (which is written in haskell)
https://getnikola.com/handbook.html#php
https://getnikola.com/handbook.html#supported-input-formats

for example: nikola static website generator(in python) has this method of enabling any markup(written in any language) in the world to be implemented as plugin with small configuration and small python wrapping
http://ralsina.me/weblog/posts/adding-support-for-a-markup-to-nikola.html

# 

so it is just a matter of providing a plugin method which will call any compatible wrapper(to be written in haxe) in your butterfly, to get any markup language(to be written in any programming language) to be plugged

**Comment by zaxebo1** [comment]
so, markdown(written in haxe,say) can be your "mandatory core" included markup language of butterfly. 
BUT plugged in markup languages(written in any language) can be an "optional" for installation, which plugs through a haxe based wrapper

**Comment by ashes999** [comment]
I think I understand what you're suggesting.

I'm open to trying it.  If it's not a huge amount of effort, and if I can write enough test automation around it, I'm okay to doing this -- although I work very sparingly on Butterfly, so I don't know how much effort it will take to maintain this. Just like with GitHub, supporting it will slow me down (more code to maintain), and I'm not sure how worthwhile it is.

There are some other issues (like how do you specify meta-data tags?) that need to be resolved.

Are you interested in working on this? Can you look into writing wrappers for asciidoctor in Haxe?

**Comment by ashes999** [comment]
I did some more analysis of this. This is what I think.

The easiest solution is for someone to create a haxelib library that includes support for parsing (and converting) Asciidoctor format into HTML. The library only needs to run on `neko`.

Asciidoctor's format is very similar to markdown (i.e. it's not like XML), so adding meta-tags (like the `meta-categories: ...` line) should be okay for authors.

If I have this, it's a pretty small change to add support to the core of Butterfly, and to write enough tests that I can reasonably maintain this in the future.

@zaxebo1 are you willing to look into this? If you can commit to writing (and maintaining) a haxelib for converting Asciidoctor to HTML, I can commit to supporting it in Butterfly in the next release (v0.4) and for the forseeable future.

If, for some reason, this doesn't pan out, I already have [documented how to create a pluggable architecture in Neko](http://ashes999.github.io/learnhaxe/neko-plugin-architecture-with-reflection.html). We can create and compile separate solution and load/use it at runtime.

**Comment by ashes999** [comment]
Alternatively, it seems like there's an existing conversion method:
- Use `asciidoc` itself to convert to `DocBook` format
- Use `pandoc` to convert from `DocBook` to `markdown`

Since conversion is probably a one-time thing, and since there's no native Haxe support/library for `asciidoc` format, I think this is the quickest, easiest solution available now.

In the long-term, supporting other formats may be worthwhile (if there are existing, well-tested libraries that make it easy to convert the output to HTML).

**Comment by ashes999** [comment]
Just a reminder that I'm waiting for a response from you @zaxebo1 . 

TLDR: if you can commit to writing an AsciiDoc-to-HTML library for Haxe, I can commit to supporting it in Butterfly. Otherwise, converting to Markdown outside of Butterfly is the way to go.

**Comment by zaxebo1** [comment]
Meanwhile I went through the butterfly source code and on second thoughts, i realised that my many more goals from static site generators are too much specific/bound to my needs , which will require myself to write my own generator; rather than burdening you with my own vision. 

So, you may close this issue. Sorry, for bothering you.

**Comment by ashes999** [comment]
No worries. If you end up writing an asciidoc Haxe extension, please drop me a comment here, and I'll re-open this issue to add support to Butterfly.

**Comment by zaxebo1** [comment]
if some day, you want asciidoctor to html convertor 'extension' in haxe, what will be the haxe interface ( http://haxe.org/manual/types-interfaces.html ) which you will want it to implement?

**Comment by ashes999** [comment]
As it stands today, the interface would include only one method:

```
public function markupToHtml(markup:String):String {
  // return html
}
```

This is [what we do today with MarkDown](https://github.com/ashes999/butterfly/blob/master/src/butterfly/core/Content.hx#L68).

If you (or anyone else) ever comes back to doing this, I will certainly create an interface for it.

I looked up the AsciiDoc syntax, and there are lots of options. It's definitely a large amount of work.
